### PR TITLE
to avoid OOM, add/update/dirty.. are processed partially

### PR DESCRIPTION
**Pull request by camparijet**
Hi, this PR is based on @hssm's suggestion in #4482 (https://github.com/ankidroid/Anki-Android/pull/4482#issuecomment-257578844)

This allows Ankidroid to load even 200k-notes-deck without any problem.

I've tested that the large test apkg (please see #4481) are successfully imported.

### CODE REVIEW

**Comment by timrae** [comment]
Thanks for this, I'll leave code review to @hssm

File: AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java [hunk 1]
```diff
@@ -173,8 +173,14 @@ private void _importNotes() {
         mIgnoredGuids = new HashMap<>();
         // iterate over source collection
         ArrayList<Object[]> add = new ArrayList<>();
+        int totalAddSize = 0;
+        final int thresExecAdd = 30000;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java]: How was the constant determined? Is it sufficiently small for all devices that OOM on large notes? Just curious - also should it be externalized into a constant at the top of the class?

File: AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java [hunk 2]
```diff
@@ -235,6 +243,31 @@ private void _importNotes() {
                     }
                 }
                 i++;
+
+                // add to col partially, so as to avoid OOM
+                if (add.size() > thresExecAdd){
+                    totalAddSize  += add.size();
+                    mDst.getDb().executeMany("insert or replace into notes values (?,?,?,?,?,?,?,?,?,?,?)", add);
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java]: These database statements are duplicated between the ones in the loop and the final one now - should they be put in an extracted method? Or is it possible to get the loop to also execute them (by checking to see if we're on the last card and executing also, perhaps?)

File: AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java [hunk 3]
```diff
@@ -248,9 +281,15 @@ private void _importNotes() {
                 cur.close();
             }
         }
+
+        // summarize partial add/update/dirty results for total values
+        totalAddSize = totalAddSize == 0 ? add.size() : add.size() + totalAddSize;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java]: is this the same as saying "totalAddSize += add.size();"? Seems like if totalAddSize is 0, then simply adding add.size() to it is correct, at which point the test isn't necessary. I could be missing something though!

File: AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java (hunk 2 above)
Comment (camparijet) [reply AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java]: I don't get your point yet, could  you elaborate what "an extracted method" ?

File: AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java (hunk 3 above)
Comment (camparijet) [reply AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java]: > is this the same as saying "totalAddSize += add.size();"? 

Thanks, you are right. All of them is equal to "totalXsize += x.size()"

**Comment by camparijet** [comment]
@mikehardy thank you for feedback. I will be happy if you can reply second one.

File: AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java (hunk 2 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java]: @camparijet sure thing! What I mean is the "extract method" refactor - my fear (which may be overdone) is that the database may change in the future and with two insert statements one of them may get fixed and one not. To guard against that, you could either pull the database work out into an extracted /separate method and just call it from both locations, or (if this is possible, I am not sure) you could maybe make the final insert part of the normal loop logic somehow.

As I said, I am never sure if the things I fear about problems with future maintenance will happen or not, but when I saw the two insert statements, the first thing I thought was "extract method"

Hope this makes sense! Thanks

Comment (camparijet) [reply AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java]: thanks @mikehardy, I understand your idea and actually that does make sense.

**Comment by mikehardy** [comment]
I looked into this pretty deeply, motivated by a desire to learn more about Android app performance and profiling.

I learned that even with the proposed changes, it is not possible to import this deck on any of the emulators I have. In fact, even if I dropped the size of the intermediate commits drastically AnkiDroid eventually ran out of JVM heap.

Furthermore, it looks like there is some very inefficient storage space usage, such that files are copied multiple times - it may be possible to reduce this.

So it is clear to me that importing collections *at some size* becomes a problem. What I am curious about now is exactly how many resources are needed at various collection sizes to do the import, in order to determine if it is a serious problem or not.  I have a hunch that with newer cheap phones and reasonable size decks (like the popular 20,000 Spanish words deck I started with) this may be important but I can't tell.

In order to move this forward I've proposed adding logging for unit tests #4863 and a small profiling module #4864 so that we can measure things and make good decisions / provably good changes

**Comment by camparijet** [comment]
Thank you for your interest, I don't think it's practical to keep this open.

**Comment by mikehardy** [comment]
No unfortunately - I really really wanted that one to work, and was kind of shocked when the testing didn't show it worked. It is still a problem (OOM on import of large decks), hopefully someone can figure something out in the future. Thanks for the attempt @camparijet

**Pull request closed without merging**